src/kernels/MIOpenAdam.cpp

```diff
@@ -0,0 +1,226 @@
+                                      float weight_decay,
+                                      float eps,
+                                      int step,
+                                      bool amsgrad,
+                                      bool maximize,
+                                      long input_size)
+{
+    size_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+    size_t gsz = gridDim.x * blockDim.x;
+
+    for(; gid < input_size; gid += gsz)
+    {
+        CTYPE grad = static_cast<CTYPE>(grad_in[gid]);
+
+        AdamInternal<PTYPE, CTYPE>(param_in,
+                                   param_out,
+                                   exp_avg_in,
+                                   exp_avg_out,
+                                   exp_avg_sq_in,
+                                   exp_avg_sq_out,
+                                   max_exp_avg_sq_in,
+                                   max_exp_avg_sq_out,
+                                   grad,
+                                   lr,
+                                   beta1,
+                                   beta2,
+                                   weight_decay,
+                                   eps,
+                                   step,
+                                   amsgrad,
+                                   maximize,
+                                   gid);
+    }
+}
+
+extern "C" __global__ void AmpAdamPacked(PTYPE* param_in,
+                                         PTYPE* param_out,
+                                         half* param_out_fp16,
+                                         GTYPE* grad_in,
+                                         PTYPE* exp_avg_in,
+                                         PTYPE* exp_avg_out,
+                                         PTYPE* exp_avg_sq_in,
+                                         PTYPE* exp_avg_sq_out,
+                                         PTYPE* max_exp_avg_sq_in,
+                                         PTYPE* max_exp_avg_sq_out,
+                                         int32_t* grad_scale,
+                                         bool* found_inf,
+                                         int* step,
+                                         float lr,
+                                         float beta1,
+                                         float beta2,
+                                         float weight_decay,
+                                         float eps,
+                                         bool amsgrad,
+                                         bool maximize,
+                                         long input_size)
+{
+    size_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+    size_t gsz = gridDim.x * blockDim.x;
+    size_t lid = threadIdx.x;
+
+    if(gid >= input_size)
+        return;
+
+    __shared__ int step_val;
+    __shared__ bool skip;
+    __shared__ float scale_factor;
+
+    if(lid == 0)
+    {
+        skip         = (found_inf) ? *found_inf : false;
```

Comment:
Who sets `found_inf`?

Reply:
Here is where found_inf is set.
[https://github.com/ROCm/pytorch/torch/amp/grad_scaler.py#L668](https://github.com/ROCm/pytorch/blob/90e886aa6cd16011a6f4fe29707e88081d712594/torch/amp/grad_scaler.py#L668)

found_inf is checked before calling the optimizer step.
[https://github.com/ROCm/pytorch/torch/amp/grad_scaler.py#L350](https://github.com/ROCm/pytorch/blob/90e886aa6cd16011a6f4fe29707e88081d712594/torch/amp/grad_scaler.py#L350)
At this point, data transfer occurs from device to host. 
Changing found_inf to be checked on the GPU prevents data transfer.

Fused Adam in ROCm PyTorch is also the same.
[https://github.com/ROCm/pytorch/blob/torch/optim/adam.py#L657](https://github.com/ROCm/pytorch/blob/90e886aa6cd16011a6f4fe29707e88081d712594/torch/optim/adam.py#L657)

Reply:
Ah, it's an external parameter. Ok.